semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
+                    | ApplyAllNoMatchError
+                    | ApplyAllAmbiguousSlotIntervalError
+  deriving (Eq,Ord,Show)
+
+
+-- | Apply a list of Inputs to the contract
+applyAllInputs :: Environment -> State -> Contract -> [Input] -> ApplyAllResult
+applyAllInputs env state contract inputs = let
+    applyAllAux
+        :: Environment
+        -> State
+        -> Contract
+        -> [Input]
+        -> [ReduceWarning]
+        -> [Payment]
+        -> ApplyAllResult
+    applyAllAux env state contract inputs warnings effects =
+        case reduceContractUntilQuiescent env state contract of
+            ReduceAllAmbiguousSlotIntervalError -> ApplyAllAmbiguousSlotIntervalError
+            ReduceAllSuccess warns effs curState cont -> case inputs of
+                [] -> ApplyAllSuccess (warnings ++ warns) (effects ++ effs) curState cont
+                (input : rest) -> case applyInput env curState input cont of
+                    Applied newState cont ->
+                        applyAllAux env newState cont rest (warnings ++ warns) (effects ++ effs)
+                    ApplyNoMatchError -> ApplyAllNoMatchError
+    in applyAllAux env state contract inputs [] []
 
 -- PROCESS
 
--- List of signatures needed by a transaction
-type TransactionSignatures = Set Party
-
-data ProcessError = PEReduceError ReduceError
-                  | PEApplyError ApplyError
+data ProcessError = PEAmbiguousSlotIntervalError
+                  | PEApplyNoMatchError
                   | PEIntervalError IntervalError
                   | PEUselessTransaction
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
-type ProcessWarning = ReduceWarning
-type ProcessEffect = ReduceEffect
 
-data ProcessResult = Processed [ProcessWarning]
-                               [ProcessEffect]
-                               TransactionSignatures
+data ProcessResult = Processed [ReduceWarning]
+                               [Payment]
                                TransactionOutcomes
                                State
                                Contract
                    | ProcessError ProcessError
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
 data Transaction = Transaction { txInterval :: SlotInterval
                                , txInputs   :: [Input] }
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
+
+type TransactionOutcomes = Map Party Money
+
 
+emptyOutcome :: TransactionOutcomes
+emptyOutcome = Map.empty
+
+
+isEmptyOutcome :: TransactionOutcomes -> Bool
+isEmptyOutcome trOut = all (== Lovelace 0) trOut
+
+
+-- Adds a value to the map of outcomes
+addOutcome :: Party -> Money -> TransactionOutcomes -> TransactionOutcomes
+addOutcome party diffValue trOut = let
+    newValue = case Map.lookup party trOut of
+        Just value -> Lovelace (getLovelace value + getLovelace diffValue)
+        Nothing    -> diffValue
+    in Map.insert party newValue trOut
```

Comment:
This is `Map.alter`

Reply:
Tried it in PS, debatable as to which is clearer:
```
-- Adds a value to the map of outcomes
addOutcome :: Party -> Money -> TransactionOutcomes -> TransactionOutcomes
addOutcome party diffValue trOut =
  let
    f (Just value) = Just (value + diffValue)

    f Nothing = Just diffValue
  in
    Map.alter f party trOut
```

Reply:
Plutus Maps don't have `alter` implemented. The simpler interface we depend on the better. I'd vote for lookup/insert.

Reply:
makes sense